nutils/matrix/__init__.py

```diff
@@ -23,27 +23,87 @@ def backend(matrix: str = 'auto'):
     return importlib.import_module('._'+matrix.lower(), __name__)
 
 
+def assemble_csr(values, rowptr, colidx, ncols):
+    '''Create sparse matrix from CSR sparse data.
+
+    Parameters
+    ----------
+    values
+        Matrix values in the row and column positions of the subsequent
+        parameters; one dimensional array of any data type.
+    rowptr
+        Compressed row indices; one dimensonal integer array of a length one up
+        from the number of matrix rows.
+    colidx
+        Column indices; one dimensional integer array of the same length as the
+        values parameter.
+    ncols
+        Number of matrix columns.
+    '''
+
+    values = numpy.asarray(values)
+    rowptr = numpy.asarray(rowptr)
+    colidx = numpy.asarray(colidx)
+    ncols = ncols.__index__()
+    if not values.ndim == 1:
+        raise MatrixError('assemble received invalid values')
+    if not (rowptr.ndim == 1 and
+            rowptr.dtype.kind in 'ui' and
+            rowptr[0] == 0 and
+            all(rowptr[1:] >= rowptr[:-1]) and
+            rowptr[-1] == len(values)):
+        raise MatrixError('assemble received invalid row indices')
+    if not (colidx.ndim == 1 and
+            colidx.dtype.kind in 'ui' and
+            len(colidx) == len(values) and
+            all(all(colidx[i+1:j] >= colidx[i:j-1]) for i, j in zip(rowptr, rowptr[1:]) if i != j) and
```

Comment:
I haven't benchmarked this line, but have you considered something along these lines (unchecked)?

```python
colidx_is_increasing = numpy.empty((len(colidx)+1,), bool)
numpy.greater_equal(colidx[1:], colidx[:-1], out=colidx_is_increasing[1:])
colidx_is_increasing[rowptr] = True
if not (colidx_is_increasing.all()):
    raise MatrixError('assemble received invalid column indices')
```

Reply:
I considered that, but favoured the conciseness of the inline test, which also has fewer comparisons. I get that the proposed form benefits from fewer instructions but doubt the difference is large enough to merit the added code complexity.

Reply:
Benchmark of the first call to `assemble_csr` in script `example/finitestrain.py`: 554 μs vs 62.1 μs. Zooming in on the specific test: 498 μs vs 8.25 μs.

Reply:
Ok fine :)